vllm/engine/llm_engine.py

```diff
@@ -344,29 +413,81 @@ def _verify_args(self) -> None:
             self.lora_config.verify_with_scheduler_config(
                 self.scheduler_config)
 
-    def encode_request(
+    def _add_processed_request(
         self,
-        request_id: str,  # pylint: disable=unused-argument
-        prompt: Optional[str],
-        prompt_token_ids: Optional[List[int]] = None,
+        request_id: str,
+        processed_inputs: LLMInputs,
+        params: Union[SamplingParams, PoolingParams],
+        arrival_time: float,
+        lora_request: Optional[LoRARequest],
+    ) -> None:
+        # Create the sequences.
+        block_size = self.cache_config.block_size
+        seq_id = next(self.seq_counter)
+        eos_token_id = None
+        if self.tokenizer:
+            eos_token_id = self.tokenizer.get_lora_tokenizer(
+                lora_request).eos_token_id
+        else:
+            logger.warning("Use None for EOS token id because tokenizer is "
```

Comment:
I'm not sure that we want to log a warning on every request. This will be a lot of log output for the no-tokenizer cases.

Reply:
Actually this came from #3748, this PR only renames the function.

Reply:
Hmm - this is probably an oversight from me. I thought this warning was added at engine initialization time, not inference time. IMO we should move it to the initialization time for sure.

Reply:
It may not be a good idea to warn during initialization time if the user indeed intends the tokenizer to not be used.

I have updated the code so that the warning still occurs on request but is only emitted at most once during the lifetime of the object.

Reply:
> It may not be a good idea to warn during initialization time if the user indeed intends the tokenizer to not be used.

I'm not sure that I understand this, I think a single warning is ok in this case, it's only a warning. And a single warning will still be logged with this latest change right? Just at the time of the first request rather than llm engine construction

Reply:
> > It may not be a good idea to warn during initialization time if the user indeed intends the tokenizer to not be used.
> 
> I'm not sure that I understand this, I think a single warning is ok in this case, it's only a warning. And a single warning will still be logged with this latest change right? Just at the time of the first request rather than llm engine construction

If the user didn't intend to use the tokenizer, then it might cause some confusion during startup. True that it's not a big deal either way though.

Reply:
I personally vote for moving this to engine initialization time since that will make our codebase cleaner. (no need for another attribute of engine just for warning count)

Reply:
Upon inspection of the [docs](https://docs.python.org/3/library/warnings.html), it seems that avoiding repeated errors is actually unnecessary:

> Warning messages are normally written to [sys.stderr](https://docs.python.org/3/library/sys.html#sys.stderr), but their disposition can be changed flexibly, from ignoring all warnings to turning them into exceptions. The disposition of warnings can vary based on the [warning category](https://docs.python.org/3/library/warnings.html#warning-categories), the text of the warning message, and the source location where it is issued. Repetitions of a particular warning for the same source location are typically suppressed.